demos/sample_azure_iot/sample_azure_iot.c

```diff
@@ -0,0 +1,625 @@
+        xTransport.pxNetworkContext = &xNetworkContext;
+        xTransport.xSend = TLS_Socket_Send;
+        xTransport.xRecv = TLS_Socket_Recv;
+
+        /* Init IoT Hub option */
+        xResult = AzureIoTHubClient_OptionsInit( &xHubOptions );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        xHubOptions.pucModuleID = ( const uint8_t * ) democonfigMODULE_ID;
+        xHubOptions.ulModuleIDLength = sizeof( democonfigMODULE_ID ) - 1;
+
+        xResult = AzureIoTHubClient_Init( &xAzureIoTHubClient,
+                                          pucIotHubHostname, pulIothubHostnameLength,
+                                          pucIotHubDeviceId, pulIothubDeviceIdLength,
+                                          &xHubOptions,
+                                          ucSharedBuffer, sizeof( ucSharedBuffer ),
+                                          ullGetUnixTime,
+                                          &xTransport );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        #ifdef democonfigDEVICE_SYMMETRIC_KEY
+            xResult = AzureIoTHubClient_SetSymmetricKey( &xAzureIoTHubClient,
+                                                         ( const uint8_t * ) democonfigDEVICE_SYMMETRIC_KEY,
+                                                         sizeof( democonfigDEVICE_SYMMETRIC_KEY ) - 1,
+                                                         Crypto_HMAC );
+            configASSERT( xResult == eAzureIoTHubClientSuccess );
+        #endif /* democonfigDEVICE_SYMMETRIC_KEY */
+
+        /* Sends an MQTT Connect packet over the already established TLS connection,
+         * and waits for connection acknowledgment (CONNACK) packet. */
+        LogInfo( ( "Creating an MQTT connection to %s.\r\n", pucIotHubHostname ) );
+
+        xResult = AzureIoTHubClient_Connect( &xAzureIoTHubClient,
+                                             false, &xSessionPresent,
+                                             sampleazureiotCONNACK_RECV_TIMEOUT_MS );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        xResult = AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xAzureIoTHubClient, prvHandleCloudMessage,
+                                                                   &xAzureIoTHubClient, sampleazureiotWAIT_FOREVER );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        xResult = AzureIoTHubClient_SubscribeDirectMethod( &xAzureIoTHubClient, prvHandleDirectMethod,
+                                                           &xAzureIoTHubClient, sampleazureiotWAIT_FOREVER );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        xResult = AzureIoTHubClient_SubscribeDeviceTwin( &xAzureIoTHubClient, prvHandleDeviceTwinMessage,
+                                                         &xAzureIoTHubClient, sampleazureiotWAIT_FOREVER );
+        configASSERT( xResult == eAzureIoTHubClientSuccess );
+
+        /* Get the device twin on boot */
```

Comment:
Nit "on boot" should probably be "after initial connection" since we don't know when sample would run in practice.

Reply:
@danewalton  @hihigupt  - another API feedback now that I'm staring at samples more.  I think `AzureIoTHubClient_GetDeviceTwin` is over-promising.  Becaues we don't actually get the twin in this call.

It's kinda more like `AzureIoTHubClient_RequestDeviceTwinAsync` (don't know about `Async` bit, in IoT C SDK it'd be 100% required).  ?

Reply:
how about AzureIoTHubClient_RequestDeviceTwin name?

Reply:
Updated comment and created issue for naming api
https://github.com/Azure/azure-iot-middleware-freertos/issues/127